### PR TITLE
Restore service port validation compatibility with 1.0/1.1

### LINKED ISSUE #21682: Headless services fail validation if targetPort is set
In 1.0 and 1.1, headless services that specified a mismatched port/targetPort were tolerated (the targetPort was ignored).

https://github.com/kubernetes/kubernetes/pull/17862#issuecomment-186666199 tightened validation to reject these services. That causes problems when upgrading from currently-working clusters.

### PR DESCRIPTION
**Pull request by liggitt**
Fixes #21682
Backs out validation change from https://github.com/kubernetes/kubernetes/pull/17862

This is breaking previously accepted and running services that incorrectly specified a targetPort in a headless service. I agree we want to warn/validate this, but we should not break services that are currently working by tightening validation on ignored values.

This is a good use case for per-version validation, where we could tighten validation in a newer version, and fixup old ignored data in conversion from the old version.

### CODE REVIEW

**Comment by liggitt** [comment]
@feihujiang @thockin PTAL

**Comment by thockin** [comment]
Can you update the comment in v1/types.go to say that targetPort has no meaning if `clusterIP` is "None" (and then regenerate the swagger and API docs, sigh)

**Comment by liggitt** [comment]
Will do

**Comment by liggitt** [comment]
@thockin something like this? (before I go regenerate everything)
```
	// This field is ignored for services with clusterIP=None, and should be omitted
	// or set equal to the service port.
```

**Comment by liggitt** [comment]
updated docs

**Comment by liggitt** [comment]
unit test error:
```
Error pulling image (0.3) from gcr.io/google_containers/kube-registry-proxy, Get https://storage.googleapis.com/artifacts.google-containers.appspot.com/containers/images/9b9342c134bbda6f57c95c52cafd4ae40333dd6cc0fce4bace9296f957042d22/ancestry: dial tcp 209.85.147.128:443: i/o timeout
```

File: pkg/api/validation/validation_test.go [hunk 1]
```diff
@@ -2626,7 +2626,10 @@ func TestValidateService(t *testing.T) {
...
 				s.Spec.ClusterIP = api.ClusterIPNone
 			},
-			numErrs: 1,
+			// in the v1 API, targetPorts on headless services were tolerated.
+			// once we have version-specific validation, we can reject this on newer API versions, but until then, we have to tolerate it for compatibility.
+			// numErrs: 1,
+			numErrs: 0,
 		},
 		{
 			name: "invalid port headless",
```
Comment (feihujiang) [inline pkg/api/validation/validation_test.go]: the above one needs to be changed to '''name: "invalid port headless 1"''',  and this changed to '''name: "invalid port headless 2"''',

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -1770,11 +1770,14 @@ func validateServicePort(sp *api.ServicePort, requireName, isHeadlessService boo
...
-		if sp.TargetPort.Type == intstr.String || (sp.TargetPort.Type == intstr.Int && sp.Port != sp.TargetPort.IntValue()) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("port"), sp.Port, "must be equal to targetPort when clusterIP = None"))
-		}
-	}
+	// in the v1 API, targetPorts on headless services were tolerated.
+	// once we have version-specific validation, we can reject this on newer API versions, but until then, we have to tolerate it for compatibility.
+	//
+	// if isHeadlessService {
+	// 	if sp.TargetPort.Type == intstr.String || (sp.TargetPort.Type == intstr.Int && sp.Port != sp.TargetPort.IntValue()) {
+	// 		allErrs = append(allErrs, field.Invalid(fldPath.Child("port"), sp.Port, "must be equal to targetPort when clusterIP = None"))
```
Comment (feihujiang) [inline pkg/api/validation/validation.go]: Needs to be changed to ```// 		allErrs = append(allErrs, field.Invalid(fldPath.Child("TargetPort"), sp.TargetPort, "must be equal to the value of the 'port' field when clusterIP = None"))```

Comment (liggitt) [inline pkg/api/validation/validation.go]: done

File: pkg/api/validation/validation_test.go (hunk 1 above)
Comment (liggitt) [inline pkg/api/validation/validation_test.go]: done

**Comment by liggitt** [comment]
all green, PTAL

File: pkg/api/validation/validation.go [hunk 3]
```diff
@@ -1770,11 +1770,14 @@ func validateServicePort(sp *api.ServicePort, requireName, isHeadlessService boo
...
-		if sp.TargetPort.Type == intstr.String || (sp.TargetPort.Type == intstr.Int && sp.Port != sp.TargetPort.IntValue()) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("port"), sp.Port, "must be equal to targetPort when clusterIP = None"))
-		}
-	}
+	// in the v1 API, targetPorts on headless services were tolerated.
+	// once we have version-specific validation, we can reject this on newer API versions, but until then, we have to tolerate it for compatibility.
+	//
+	// if isHeadlessService {
+	// 	if sp.TargetPort.Type == intstr.String || (sp.TargetPort.Type == intstr.Int && sp.Port != sp.TargetPort.IntValue()) {
+	// 		allErrs = append(allErrs, field.Invalid(fldPath.Child("targetPort"), sp.Port, "must be equal to the value of 'port' when clusterIP = None"))
```
Comment (feihujiang) [inline pkg/api/validation/validation.go]: s/sp.Port/sp.TargetPort

Comment (liggitt) [inline pkg/api/validation/validation.go]: fixed

**Comment by feihujiang** [comment]
@liggitt This PR needs rebase.

**Comment by liggitt** [comment]
just rebased

**Comment by feihujiang** [comment]
LGTM.

**Comment by thockin** [comment]
Thanks

**Comment by liggitt** [comment]
@k8s-bot test this issue: #21451

**Comment by ixdy** [comment]
@k8s-bot test this, issue #IGNORE (restarted Jenkins)

**Pull request merged**